dali/kernels/signal/fft/fft_cpu.h

```diff
@@ -0,0 +1,137 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SIGNAL_FFT_FFT_CPU_H_
+#define DALI_KERNELS_SIGNAL_FFT_FFT_CPU_H_
+
+#include <memory>
+#include <complex>
+#include "dali/core/common.h"
+#include "dali/core/error_handling.h"
+#include "dali/core/format.h"
+#include "dali/core/util.h"
+#include "dali/kernels/kernel.h"
+
+namespace dali {
+namespace kernels {
+namespace signal {
+namespace fft {
+
+enum FftSpectrumType {
+  FFT_SPECTRUM_COMPLEX = 0,    // separate interleaved real and img parts: (r0, i0, r1, i1, ...)
+  FFT_SPECTRUM_MAGNITUDE = 1,  // sqrt( real^2 + img^2 )
+  FFT_SPECTRUM_POWER = 2,      // real^2 + img^2
+};
+
+struct FftArgs {
+  FftSpectrumType spectrum_type = FFT_SPECTRUM_COMPLEX;
+  int transform_axis = -1;
+  int nfft = -1;
+
+  inline bool operator==(const FftArgs& oth) const {
+    return spectrum_type == oth.spectrum_type &&
+           transform_axis == oth.transform_axis &&
+           nfft == oth.nfft;
+  }
+
+  inline bool operator!=(const FftArgs& oth) const {
+    return !operator==(oth);
+  }
+};
+
+namespace impl {
+
+template <typename OutputType = std::complex<float>, typename InputType = float, int Dims = 2>
+class DLL_PUBLIC FftImpl {
+ public:
+  static_assert(std::is_same<InputType, float>::value,
+    "Data types other than float are not yet supported");
+
+  static_assert(std::is_same<OutputType, float>::value
+             || std::is_same<OutputType, std::complex<float>>::value,
+    "Data types other than float are not yet supported");
+
+  DLL_PUBLIC virtual ~FftImpl() = default;
+
+  DLL_PUBLIC virtual KernelRequirements Setup(KernelContext &context,
+                                              const InTensorCPU<InputType, Dims> &in,
+                                              const FftArgs &args) = 0;
+
+  DLL_PUBLIC virtual void Run(KernelContext &context,
+                              const OutTensorCPU<OutputType, Dims> &out,
+                              const InTensorCPU<InputType, Dims> &in,
+                              const FftArgs &args) = 0;
+};
+
+}  // namespace impl
+
+/**
+ * @brief Computes 1-D FFT related transformation from real data to either a complex spectrum
+ *   or a transformation of the complex spectrum (power, magnitude)
+ *
+ * Input data can be a 2D or 3D tensor representing a signal (e.g. [channels, time]) or
+ * a sequence of frames (e.g. [channels, frames, time]). The kernel can work with other data
+ * layouts by providing the transform_axis representing the dimension to be transformed to the
+ * frequency domain (e.g. for a layout of [channels, time, frames] we set transform_axis=1 to
+ * produce a [channels, frequency, frames] layout)
+ *
+ * @param args.spectrum_type defines the nature of the output
+ *   FFT_SPECTRUM_COMPLEX:
```

Comment:
All output types should produce same number of elements - for real input it'd be N/2+1, for complex inputs it would be N.

Reply:
will make it `n/2+1` for now since we don't handle complex inputs